compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,409 @@
+                ty::Int(IntTy::I128) => "u4i128",
+                ty::Int(IntTy::Isize) => "u5isize",
+                ty::Uint(UintTy::U8) => "u2u8",
+                ty::Uint(UintTy::U16) => "u3u16",
+                ty::Uint(UintTy::U32) => "u3u32",
+                ty::Uint(UintTy::U64) => "u3u64",
+                ty::Uint(UintTy::U128) => "u4u128",
+                ty::Uint(UintTy::Usize) => "u5usize",
+                ty::Float(FloatTy::F32) => "u3f32",
+                ty::Float(FloatTy::F64) => "u3f64",
+                _ => "",
+            });
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Char => {
+            // u<type-name> as vendor extended type
+            let mut s = String::from("u4char");
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Str => {
+            // u<type-name> as vendor extended type
+            let mut s = String::from("u3str");
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Never => {
+            // u<type-name> as vendor extended type
+            let mut s = String::from("u5never");
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Compound types
+        // () in Rust is equivalent to void return type in C
+        ty::Tuple(_) if ty.is_unit() => {}
+
+        // Sequence types
+        ty::Tuple(tys) => {
+            // u5tupleI<element-type1..element-typeN>E as vendor extended type
+            let mut s = String::from("u5tupleI");
+            for ty in tys.iter() {
+                s.push_str(&encode_ty(tcx, ty, dict, encode_c_types));
+            }
+            s.push('E');
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Array(ty0, len) => {
+            // u5arrayI<array-length><element-type>E as vendor extended type
+            let mut s = String::from("u5arrayI");
+            s.push_str(
+                format!("{}", len.val().try_to_scalar().unwrap().to_i64().unwrap()).as_ref(),
+            );
+            s.push_str(&encode_ty(tcx, *ty0, dict, encode_c_types));
+            s.push('E');
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Slice(ty0) => {
+            // u5sliceI<element-type>E as vendor extended type
+            let mut s = String::from("u5sliceI");
+            s.push_str(&encode_ty(tcx, *ty0, dict, encode_c_types));
+            s.push('E');
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        // User-defined types
+        // FIXME(rcvalle): Identify C char and integer type uses and encode them with their
+        // respective builtin type encoding as specified by the Itanium C++ ABI if encode_c_types
+        // is true.
+        ty::Adt(adt_def, ..) => {
+            // S<length><name>
```

Comment:
This production doesn't use terminal `S`?

Reply:
I think I may have misunderstood the examples at https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling.named.

These functions:
```
struct struct1 *
func_struct11(struct struct1 *arg)
{
    printf("%s: %p received\n", __func__, arg);
    return arg;
}

struct struct1
func_struct12(struct struct1 arg)
{
    printf("%s: %p received\n", __func__, &arg);
    return arg;
}
```

Are encoded by Clang as:
```
!149 = !{i64 0, !"_ZTSFP7struct1S0_E"}
!150 = !{i64 0, !"_ZTSFPvS_E.generalized"}
...
!160 = !{i64 0, !"_ZTSF7struct1S_E"}
!161 = !{i64 0, !"_ZTSF7struct1S_E.generalized"}
```

So it seems it's just <length><name> (i.e., a simple <unqualified-name> and there is no 'S')?

Reply:
In the example `1S` encodes the name of `struct S`.

Reply:
Looking at the code again, it seems I did the encoding correctly, but confusedly put the 'S' in the comments and the design doc for some reason. Could you confirm if the encoding is correct or if there is anything I need to fix besides the comments?

Reply:
Yes, encoding seems fine. I was just referring to the `S` in the comment.

Same thing applies to comment in `ty::Foreign`.

Reply:
Done.